### DISCUSSION TITLE
Want to run a SH1107 Display with Platform.io and Rp2040 over I²C (dont work)

### DISCUSSION THREAD

**Discussion by Feliniuss** [olikraus/u8g2#2345] [Q&A]
hello there, I would like to start by saying that I am a complete beginner. At best, if that is possible, I would like to display variables as numerical values, i.e. a voltage and current measurement with a display. For this I use a RP2040 with which I want to control the display of Waveshare ( https://www.waveshare.com/wiki/Pico-OLED-1.3#I2C ) via i2C. If I have understood the library correctly, this should be relatively simple. I have already tried the Hello world example but there was no feed back. In the code I used the following constructor: U8G2_SH1107_64X128_1_HW_I2C u8g2(U8G2_R0, U8X8_PIN_NONE, 17, 16); 

16 17 for SDA and SCL because i built my whole pcb by my selfe so it probably can be possible that i have a hardware bug but i know i2C works and when i pull up with this code 

pinMode(9,OUTPUT);//RESET 
digitalWrite(9, LOW);
pinMode(7,OUTPUT);//Chipselect
digitalWrite(7, LOW);
digitalWrite(9, HIGH);

 CS and Reset is correct so that i can read the i2C Adresses in the "NET" and i get the normal 0x03C (60dec) address aknowlage.

if i have forgotten any information that might be needed let me know i would be extremely grateful if someone could explain to me what is missing, could be missing or in principle can not work with the program.

**Comment by olikraus** [comment]
I assume that the RP2040 doesn't support the pin remapping as used in your example. I suggest to start with SW emulated I2C first, to ensure that there are no hw bugs:
U8G2_SH1107_64X128_1_SW_I2C u8g2(U8G2_R0, 17, 16, U8X8_PIN_NONE);

**Comment by Feliniuss** [comment]
i don't quite understand your thoughts but i will try tomorrow with the constructor thanks for the quick feedback. from the hardware point of view i can't imagine that something is wrong because the rest of my design works now i want to output the sensor data to the display.

**Comment by Feliniuss** [comment]
ok i have tried it now with the constructor unfortunately nothing happens here what exactly is the difference between software and hardware i2C. and do i understand correctly that i have connected the Rp2040 to 2 other Gpios is possibly a problem now? what exactly does the reset pin for the display mean correctly that is with me on Gpio 9 which i set to 1 via digitalWrite(9, HIGH); to switch on the display so i can then also read out my I2C address.

#include <Arduino.h>
#include <U8g2lib.h>

#ifdef U8X8_HAVE_HW_SPI
#include <SPI.h>
#endif
#ifdef U8X8_HAVE_HW_I2C
#include <Wire.h>
#endif
//#define SCREEN_ADDRESS 0x3D ///< See datasheet for Address; 0x3D for 128x64, 0x3C for 128x32

/*
  U8g2lib Example Overview:
    Frame Buffer Examples: clearBuffer/sendBuffer. Fast, but may not work with all Arduino boards because of RAM consumption
    Page Buffer Examples: firstPage/nextPage. Less RAM usage, should work with all Arduino boards.
    U8x8 Text Only Example: No RAM usage, direct communication with display controller. No graphics, 8x8 Text only.
    
  This is a page buffer example.    
*/

// Please UNCOMMENT one of the contructor lines below
// U8g2 Contructor List (Picture Loop Page Buffer)
// The complete list is available here: https://github.com/olikraus/u8g2/wiki/u8g2setupcpp
// Please update the pin numbers according to your setup. Use U8X8_PIN_NONE if the reset pin is not connected


//U8G2_SH1107_64X128_1_HW_I2C u8g2(U8G2_R0, U8X8_PIN_NONE, 17, 16);
//U8G2_SH1107_64X128_1_2ND_HW_I2C u8g2(U8G2_R0, U8X8_PIN_NONE);
U8G2_SH1107_64X128_1_SW_I2C u8g2(U8G2_R0, 17, 16, U8X8_PIN_NONE);


// End of constructor list


void setup(void) {
    Serial.begin(115200);
  pinMode(9,OUTPUT);//RESET 
digitalWrite(9, LOW);
pinMode(7,OUTPUT);//Chipselect
digitalWrite(7, LOW);
digitalWrite(9, HIGH);
//u8g2.setI2CAddress(SCREEN_ADDRESS*2);
	

  u8g2.begin();  
}

void loop(void) {
  u8g2.firstPage();
  do {
  
    u8g2.drawHLine(1,1,10);
    u8g2.drawHLine(1+5,2,5);
  
    u8g2.setFont(u8g2_font_ncenB10_tr);
    u8g2.drawStr(2,2,"Hello World!");
  } while ( u8g2.nextPage() );
  delay(1000);
}

**Reply by olikraus** [reply]
Will your example work?
Somehow your code is confusing. I2C does not have a CS pin!
Are you sure, you have a I2C display? You might need to use the SPI constructor if your display has an CS pin.

**Reply by Feliniuss** [reply]
Hi, I think I'll explain it in a bit more detail. I made the hardware for my project myself (Rp2040) with 6 I2C participants. Now I wasn't sure if cs and reset are important for I2C or actually for the Waveshare 1.3 OLED, so I connected them. now I just want to set it to 0 like in the code CS and pull reset to 1 so that the display is "on". in the meantime I got it to work with another library. I would still be interested to know if it works with yours. However, I have not found any example code that gives me at least a small sign of life on the display. And as a beginner, I personally find it difficult to interpret in code how to write what. Maybe there is a way to control it, I just couldn't figure it out.

**Reply by olikraus** [reply]
Maybe you can provide the unit code and pin assignments for the other lib.

**Reply by Feliniuss** [reply]
sorry for the late reply. I had used the OneBitDisplay library: 
ONE_BIT_DISPLAY obd;
obd.setI2CPins(16, 17); //SDA ,SCL
and the display address 0x3C is automatically Detected.
Is that what you mean? Did I understand you correctly ?
Now I would like to get the display running with your library.

**Reply by olikraus** [reply]
hmm one but display lib does not look like a regular Arduino lib, so maybe the pin numbers are different.
Use the U8G2_SH1107_64X128_1_SW_I2C u8g2(U8G2_R0, 17, 16, U8X8_PIN_NONE); constructor, but probably the pin numbers are different.
According to this picture: 
![pinout](https://github.com/olikraus/u8g2/assets/1084992/cbda9110-969a-4a98-af8a-89839db80c7e)

the correct Arduino pins could be 5 & 4: `U8G2_SH1107_64X128_1_SW_I2C u8g2(U8G2_R0, 5, 4, U8X8_PIN_NONE);`

**Comment by mustafaoral** [comment]
I'm using u8x8 instead of u8g2, a RPi Pico and a similar display from Waveshare ([SKU 18179](https://www.waveshare.com/product/displays/oled/oled-3/1.3inch-oled-module-c.htm)) which uses SH1107. I had to the connect reset cable to an unused GPIO pin and set that as the reset pin in the constructor like `U8X8_SH1107_64X128_HW_I2C(18, 1, 0);`. Not only that, I had to pull DC low. 

Why all that should be necessary for a supposed I2C display that should use four cables, that I don't know, but hope this helps.
